js/plugins/api.js

```diff
@@ -0,0 +1,213 @@
+import config from 'config';
+import log from 'logger';
+
+
+export class ApiError extends Error {
+    constructor(response, data) {
+        const message = data.hasOwnProperty('message')
+            ? data.message
+            : response.statusText || 'API Error';
+
+        super(message);
+
+        this.response = response;
+        this.status = response.status;
+        this.data = data;
+    }
+}
+
+
+/**
+ * Expect JSON by default
+ */
+const DEFAULT_HEADERS = {'Accept': 'application/json'};
+
+/**
+ * Write operations need some extras headers:
+ * Send JSON by default and give the x-csrf token
+ *
+ * Credtentials might be needed here but sessions is used
+ */
+const WRITE_HEADERS = {
+    'Content-Type': 'application/json',
+    'X-CSRFToken': config.csrftoken,
+};
+
+
+/**
+ * Wrap fetch() calls with predefined API parameters
+ */
+export class Api {
+    constructor(root) {
+        // Compute the root once
+        root = root || '';
+        this.root = root.endsWith('/')
+            ? root.substring(0, root.length - 1)
+            : root;
+        // Expose ApiError as Vue.api.Error
+        this.Error = ApiError;
+    }
+
+    /**
+     * Tells wether or not a given URL is absolute.
+     * @param  {String}  url The URL to test
+     * @return {Boolean}     True is the URL is absolute
+     */
+    isAbsolute(url) {
+        return /^https?:\/\//.test(url) || url.startsWith('/');
+    }
+
+    /**
+     * Encode paramaters as Object to querystring
+     * @param  {Object} params The parameters to encode
+     * @return {String}        The URL encoded querystring
+     */
+    encodeParams(params) {
+        return Object.keys(params).map(key => {
+            return [key, params[key]].map(encodeURIComponent).join('=');
+        }).join('&');
+    }
+
+    /**
+     * Build a complete URL
+     * @param  {String} url     An absolute or API root relative URL
+     * @param  {Object} params  Optionnal query strings parameters
+     * @return {String}         The full URL
+     */
+    build(url, params) {
+        if (!this.isAbsolute(url)) {
+            url  = `${this.root}/${url}`;
+        }
+        if (params && Object.keys(params).length) {
+            const parts = [url, this.encodeParams(params)];
+            url = url.indexOf('?') >= 0 ? parts.join('&') : parts.join('?');
+        }
+        return url;
+    }
+
+    /**
+     * Default handler for JSON response
+     * Parse the response if OK otherwise throw it
+     * (let the developper handle it in the catch() handler)
+     * @param  {Response} response A response with JSON expected body.
+     * @return {Promise}           A promise for data handling
+     */
+    onResponse(response) {
+        if (response.ok) {
```

Comment:
Yes:
https://developer.mozilla.org/en-US/docs/Web/API/Response/ok
https://github.com/github/fetch/pull/82
And there is a test case.